### PR TITLE
ao_wasapi: fix player core lockup when avoiding premature buffer fills

### PR DESCRIPTION
**Pull request by na-na-hi**
https://github.com/mpv-player/mpv/commit/6863eefc3d1d683a08f56aeaca0b4706672f2447 handled this situation by using an atomic variable to express the state for which the wakeup is caused by AO control, and the dispatch queue is only processed at this state. However, this can cause permanent lockup of the player core when the following happens:

- AO control sets the thread state to `WASAPI_THREAD_DISPATCH`, and sets the wakeup handle.
- WASAPI thread reads the `WASAPI_THREAD_DISPATCH` state and processes the dispatch queue.
- Another AO control happens. A dispatch item is enqueued, and the state stays at `WASAPI_THREAD_DISPATCH`.
- WASAPI thread resets the thread state to `WASAPI_THREAD_FEED` since the state has not changed.
- `WaitForSingleObject()` returns in the WASAPI thread, sees this state, and does not process the dispatch queue.
- The player core locks permanently because it is waiting for the dispatch to be processed.

This has been experimentally verified on a system under high contention: The easiest way to trigger this lockup is to continuously hold down "i", which rapidly issues AO get volume/mute controls.

To properly handle this, do not skip the dispatch queue processing. Instead, set a flag which indicates that the thread is woke up due to AO control. When this happens, skip the rest of WASAPI loop processing.

Fixes: https://github.com/mpv-player/mpv/commit/6863eefc3d1d683a08f56aeaca0b4706672f2447

### CODE REVIEW

**Comment by kasper93** [comment]
Can you please use `WaitForMultipleObjects`, thank you. You can see related discussion https://github.com/mpv-player/mpv/pull/13687

**Comment by kasper93** [comment]
> Is there any reason why we can't use the dispatch queue for reset/resume/set_pause requests?

I don't know, but the main thing as I remember reading it last time was that we didn't know if the thread was woken up externally for feed or by us for thread state or dispatch msg. If we had separate events for feed and everything else, we can use `WaitForMultipleObjects` and check what reason was for our wake up.

**Review by kasper93 — APPROVED** [review]

**Comment by kasper93** [comment]
@sunpenghao: Can you try this patch on your end?

File: audio/out/ao_wasapi.c [hunk 1]
```diff
@@ -220,18 +222,25 @@ static DWORD __stdcall AudioThread(void *lpParameter)
...
+        switch (WaitForMultipleObjects(MP_ARRAY_SIZE(handles), handles, FALSE, INFINITE)) {
+        case WAIT_OBJECT_0:
+            // fill twice on under-full buffer (see comment in thread_feed)
+            if (thread_feed(ao) && thread_feed(ao))
+                MP_ERR(ao, "Unable to fill buffer fast enough\n");
+            continue;
+        case WAIT_OBJECT_0 + 1:
+            break;
+        default:
             MP_ERR(ao, "Unexpected return value from WaitForSingleObject\n");
```
Comment (sunpenghao) [inline audio/out/ao_wasapi.c:235]: Nit:
```suggestion
            MP_ERR(ao, "Unexpected return value from WaitForMultipleObjects\n");
```

**Review by sunpenghao — APPROVED** [review]
Works as expected. Sorry to have forgotten about this issue.

File: audio/out/ao_wasapi.c [hunk 2]
```diff
@@ -220,18 +222,25 @@ static DWORD __stdcall AudioThread(void *lpParameter)
...
     while (true) {
-        if (WaitForSingleObject(state->hWake, INFINITE) != WAIT_OBJECT_0)
-            MP_ERR(ao, "Unexpected return value from WaitForSingleObject\n");
-
-        int thread_state = atomic_load(&state->thread_state);
-        switch (thread_state) {
-        case WASAPI_THREAD_FEED:
+        HANDLE handles[] = {state->hWake, state->hUserWake};
+        switch (WaitForMultipleObjects(MP_ARRAY_SIZE(handles), handles, FALSE, INFINITE)) {
+        case WAIT_OBJECT_0:
```
Comment (kasper93) [inline audio/out/ao_wasapi.c:227]: There is one possible issue with this. If multiple objects become signaled, the function returns the index of the first handle in the array whose object was signaled. So there is possibility that we ignore dispatch wake up if both happen.

**Comment by kasper93** [comment]
Ok, fair. Make sense it works this way, otherwise it would be awkward.

**Pull request merged**